frontend/src/app/components/routing/protected-routes.tsx

```diff
@@ -1,15 +1,46 @@
 import { FC } from "react";
 import { Navigate, Outlet } from "react-router-dom";
+import jwt_decode from "jwt-decode";
 import Roles from "../../types/app/roles";
 import Layout from "../containers/layout";
+import { AUTH_TOKEN } from "../../service/user-service";
 
-export const ProtectedRoutes: FC<{ roles: Array<Roles> }> = () => {
-  let auth = { token: true };
-  return auth.token ? (
+interface ProtectedRoutesProps {
+  roles: Array<Roles>;
+}
+
+interface DecodedToken {
+  client_roles: Array<Roles>;
+}
+
+const getAuthToken = (): string | null => {
+  return localStorage.getItem(AUTH_TOKEN);
+};
+
+const hasRequiredRole = (userRoles: Array<Roles>, requiredRoles: Array<Roles>): boolean => {
+  return requiredRoles.some((role) => userRoles?.includes(role));
+};
+
+export const ProtectedRoutes: FC<ProtectedRoutesProps> = ({ roles }) => {
+  const token = getAuthToken();
+  if (!token) {
+    return <Navigate to="/not-authorized" />;
+  }
+
+  let decodedToken: DecodedToken;
```

Comment:
I like what you've done returning a `<Navigate to="/not-authorized" />;` but you can simplify it a bit more.

```
  const _hasAccess = (requiredRoles: Array<Roles>): boolean => {
    let hasAccess = true;
    const token = localStorage.getItem(AUTH_TOKEN);

    try {
      if (!token) {
        throw Error();
      }

      const { client_roles } = jwtDecode(token) as { client_roles: Array<Roles> };
      if (client_roles.length === 0) {
        hasAccess = false;
      }

      hasAccess = client_roles.some((role) => requiredRoles.includes(role));
    } catch (Error) {
      hasAccess = false;
    } finally {
      return hasAccess;
    }
  };

  if (!_hasAccess(requiredRoles)) {
    return <Navigate to="/not-authorized" />;
  }

  return (
    <Layout>
      <Outlet />
    </Layout>
  );```
  
if you want to keep things as is at the very least you'll want to move your helpers into your component

Reply:
I cleaned this up a bit, I _think_ my chages will address your concerns.